src/ui/pages/config/PageConfigGrid.js

```diff
@@ -0,0 +1,82 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+
+const renderGrid = (gridStruct, parentBsWidth) => {
+  if (!gridStruct.rows) {
+    return null;
+  }
+  return gridStruct.rows.map((row) => {
+    const renderedCols = row.cols.map((col) => {
+      const width = (col.x2 - col.x1) + 1;
+      const height = (col.y2 - col.y1) + 1;
+      const pcWidth = ((width / parentBsWidth) * 100);
+      let className;
+      if (col.rows) {
+        className = 'PageConfigGrid__container';
+      } else if (!col.frame) {
+        className = 'PageConfigGrid__hole';
+      } else {
+        className = 'PageConfigGrid__slot';
+      }
+
+      return (
+        <div
+          key={`col-${col.x1}`}
+          style={{ minHeight: 80 * height, width: `${pcWidth}%` }}
+          className={className}
+        >
+          {col.rows ? renderGrid(col, width) : col.frame && col.frame.descr }
+        </div>
+      );
+    });
+    return (
+      <div
+        key={`row-${row.y1}`}
+        className="PageConfigGrid__row"
+      >
+        {renderedCols}
+      </div>
+    );
+  });
+};
```

Comment:
lets split this into more methods it is really hard to read it in its current state